File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/teleport/TeleportOptions.java
package top.likoslupus.cellulosesz.api.teleport;

import static top.likoslupus.cellulosesz.api.validation.NumericChecks.requireNonNegative;

public record TeleportOptions(
        boolean safe,
        boolean rememberBack,
        boolean allowCrossWorld,
        boolean keepVehicle,
        int warmupSeconds
) {

    public TeleportOptions {
        requireNonNegative(warmupSeconds, "warmupSeconds");
    }

    public static TeleportOptions defaults() {
        return new TeleportOptions(true, true, true, false, 0);
    }

    public TeleportOptions withoutBackMemory() {
        return new TeleportOptions(safe, false, allowCrossWorld, keepVehicle, warmupSeconds);
    }

    public TeleportOptions withWarmup(int seconds) {
        return new TeleportOptions(safe, rememberBack, allowCrossWorld, keepVehicle, seconds);
    }

    public TeleportOptions withSafe(boolean value) {
        return new TeleportOptions(
                value,
                rememberBack,
                allowCrossWorld,
                keepVehicle,
                warmupSeconds
        );
    }

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/teleport/TeleportOptions.kt
package top.likoslupus.cellulosesz.api.teleport

import top.likoslupus.cellulosesz.api.validation.requireNonNegative

@JvmRecord
public data class TeleportOptions(
    public val safe: Boolean,
    public val rememberBack: Boolean,
    public val allowCrossWorld: Boolean,
    public val keepVehicle: Boolean,
    public val warmupSeconds: Int
) {

    init {
        warmupSeconds.requireNonNegative { "warmupSeconds" }
    }

    public fun withoutBackMemory(): TeleportOptions =
        TeleportOptions(
            safe,
            false,
            allowCrossWorld,
            keepVehicle,
            warmupSeconds
        )

    public fun withWarmup(seconds: Int): TeleportOptions =
        TeleportOptions(
            safe,
            rememberBack,
            allowCrossWorld,
            keepVehicle,
            seconds
        )

    public fun withSafe(value: Boolean): TeleportOptions =
        TeleportOptions(
            value,
            rememberBack,
            allowCrossWorld,
            keepVehicle,
            warmupSeconds
        )

    public companion object {

        @JvmStatic
        public fun defaults(): TeleportOptions =
            TeleportOptions(
                safe = true,
                rememberBack = true,
                allowCrossWorld = true,
                keepVehicle = false,
                warmupSeconds = 0
            )

    }

}
